replicator/src/main/scala/com/evolutiongaming/kafka/journal/replicator/Batch.scala

```diff
@@ -13,140 +30,103 @@ private[journal] sealed abstract class Batch extends Product {
-                if (a.to > b.to) {
-                  val cleanTail: List[Batch] = tail.mapFilter {
-                    case appends: Appends => dropDeleted(appends, a)
-                    case delete: Delete   => delete.some
-                    case purge: Purge     => purge.some
-                  }
-
-                  val delete = deleteOf(a.to, b.origin orElse a.origin, b.version orElse a.version)
-                  delete :: cleanTail
-                } else {
-                  val delete = b.copy(offset = offset, origin = b.origin orElse a.origin, version = b.version orElse a.version)
-                  delete :: tail
-                }
-
-              case (_: Delete, a: Action.Purge) =>
-                purgeOf(a.origin, a.version) :: Nil
-
-              case (b: Purge, a: Action.Append) =>
-                appendsOf(NonEmptyList.of(actionRecord(a))) :: b :: Nil
-
-              case (b: Purge, _: Action.Mark) =>
-                b.copy(offset = offset) :: Nil
-
-              case (b: Purge, a: Action.Delete) =>
-                deleteOf(a.to, a.origin, a.version) :: b :: Nil
-
-              case (_: Purge, a: Action.Purge) =>
-                purgeOf(a.origin, a.version) :: Nil
-            }
-
-          case Nil =>
-            record.action match {
-              case a: Action.Append => appendsOf(NonEmptyList.of(actionRecord(a))) :: Nil
-              case _: Action.Mark   => Nil
-              case a: Action.Delete => deleteOf(a.to, a.origin, a.version) :: Nil
-              case a: Action.Purge  => purgeOf(a.origin, a.version) :: Nil
-            }
+    def batches: List[Batch] = {
+      // we can drop first `append`, if `deleteTo` will discard it AND there is at least one more `append`
+      // we have to preserve one `append` to store latest `seqNr` and populate `expireAfter`
+      val appends = {
+        this.appends.flatMap { appends =>
+          val deleteTo = this.delete.map(_.to.value)
+          val records  = appends.records
+          val actions =
+            if (deleteTo.contains(records.head.action.range.to)) NonEmptyList.fromList(records.tail).getOrElse(records)
+            else records
+          appends.copy(records = actions).some
         }
       }
-      .foldLeft(List.empty[Batch]) { (bs, b) => // reverse order of batches
-        b match {
-          case b: Appends => b.copy(records = b.records.reverse) :: bs // reverse append actions
-          case b: Delete  => b :: bs
-          case b: Purge   => b :: bs
-        }
+
+      // if `delete` was not last action, adjust `delete`'s batch offset to update `metajournal` correctly
+      val delete = appends match {
+        case Some(appends) => this.delete.map(delete => delete.copy(offset = delete.offset max appends.offset))
+        case None          => this.delete
       }
+
+      // apply action batches in order: `Purge`, `Append`s and `Delete`
+      List(purge, appends, delete).flatten
```

Comment:
I totally agree that purge must be executed first, but I have strong doubts about doing append before delete because:

1. we're not interested in appends before delete (excluding the case of last append for propagating `seqNr` and `expiration`)
2. all (not dropped) appends actually coming _after_ delete

I propose to change the order of batches to: purge, delete, append* where delete batch will contains not only delete data but optionally last append before the delete. This will remove need in any recalculations of delete.offset, simplifies merging deletes and makes order of batches equal to original events order

Reply:
I was under impression that we do not want to change the "contract" beyond `Batch` and its users, but I can rework code according to your proposal, though then I have to change the "shape" of the `Delete`